File: mt/computer_architecture_3021/assignments/6/cache.js
class Cache {

  // L -> Bytes per cache line
  // K -> Data directories
  // N -> Sets of cache lines/tags

  constructor(L, K, N){
    // L, K, N Variables for K-Way Cache
    this.L = L;
    this.K = K;
    this.N = N;

    // Set cache tag variables
    this.offset = 4;
    this.set_selector_bits_length = Math.log2(N);
    this.size = L * K * N;
    this.tag_bits_length = L - this.offset - this.set_selector_bits_length;

    // Cache matrix
    this.cache_matrix = [];

    // Initialise cache matrix
    for(var i=0; i<=N; i++){
      this.cache_matrix.push([(i).toString(2).padStart(this.set_selector_bits_length, 0), 0])
      for(var j=0; j<=K; j++) {
        var d = new Date()
        this.cache_matrix[i].push([0,"",this.getTime()])
      }
    }
  }

  getTime(){
    var d = new Date()
    return d.getTime();
  }

  printStats(){
    console.log("L -> " + this.L);
    console.log("K -> " + this.K);
    console.log("N -> " + this.N);

    console.log("Size -> " + this.size);
    console.log("Offset -> " + this.offset);
    console.log("Set selector length (Bits) -> " + this.set_selector_bits_length);
    console.log("Tag length (Bits) -> " + this.tag_bits_length);
    //console.log("Cache Matrix -> " + this.cache_matrix.toString())
  }

  hexToBinary(hex){
    return (parseInt(hex)).toString(2).padStart(16, 0);
  }

  disectBinary(binary){
    //    <-  0-12  ->      <- 12-14 ->
    // [BINARY_MINUS_OFFSET] [OFFSET]
    var binary_minus_offset = binary.substr(0,12);
    var offset = binary.substr(12, binary.length-1);

    //    <-  0-12  ->
    // [BINARY_MINUS_OFFSET]
    // [   TAG    ][BIT_SEL]
    var bit_selector_bits = binary_minus_offset.slice(-this.set_selector_bits_length);
    var tag_bits;

    if(binary_minus_offset.length-this.set_selector_bits_length != 0){
        tag_bits = binary_minus_offset.substr(0, binary_minus_offset.length-this.set_selector_bits_length);
    }

    if(this.N == 1){
      tag_bits = bit_selector_bits;
      bit_selector_bits = '0';
    }

    return [bit_selector_bits, tag_bits, offset];
  }

  find_lru(timestamps){
    var oldest = new Date(Math.min.apply(null, timestamps));
    var index = timestamps.indexOf(oldest)
    return index
  }

  handleOneDataDirectory(tag_bits, row){
    // If hasData == 0 then miss
    if(row[2][0] == 0){
      // Return miss and pretend fetch from memory
      row[2][1] = tag_bits;
      row[2][0] = 1;
      return("MISS");
    }
    // If hasData == 1
    else if(row[2][0] == 1){
      // && If tag_bits_match()
      if(row[2][1] == tag_bits)
        return("HIT");
      else{
        // Return miss and pretend fetch from memory
        row[2][1] = tag_bits;
        return("MISS")
      }
    }
  }

  handleLRUReplacement(tag_bits, row){
    var timestamps = [];

    // Loop over all possible data directories for the 1 set
    for(var index=0; index<=this.K; index++){
      timestamps.push(row[2+index][2]);

      // If directory has data and matches tag_bits --> HIT
      if(row[2+index][0] == 1 && row[2+index][1] == tag_bits){
        // Update time for LRU
        row[2+index][2] = this.getTime();
        return("HIT")
      }
    }

    // If no HIT found, check LRU and replace
    row[1] = this.find_lru(timestamps);
    var lru = row[1]

    // If empty, simulate MISS and fetch
    if(row[2+lru][0] == 0){
      row[2+lru][0] = 1;
      row[2+lru][1] = tag_bits;
      row[2+lru][2] = this.getTime();
      return("MISS");
    }
    // Otherwise, replace entry and simulate MISS and fetch
    else if(row[2+lru][0] == 1){
      row[2+lru][1] = tag_bits;
      row[2+lru][2] = this.getTime();
      return("MISS");
    }
  }

  hitOrMiss(bit_selector_bits, tag_bits){
    // Iterate over every row in matrix
    for(var i=0; i<this.cache_matrix.length; i++){
      var row = this.cache_matrix[i]
      // If bit_selector_bits matches cache matrix
      if(bit_selector_bits == row[0]){
        // If only one data directory
        if(this.K == 1){
          //return this.handleOneDataDirectory(tag_bits, row);
        }
        // Otherwise, handle LRU replacement
        else{
          return this.handleLRUReplacement(tag_bits, row);
        }
      }
    }
  }
}

const cache = new Cache(16, 8, 1);

var input = ["0000","0004","000c","2200","00d0","00e0","1130","0028",
        		 "113c","2204","0010","0020","0004","0040","2208","0008",
        		 "00a0","0004","1104","0028","000c","0084","000c","3390",
        		 "00b0","1100","0028","0064","0070","00d0","0008","3394"]

var hits = 0;
var misses = 0;

// Test all input hex addresses
for(hex in input){

  // Generate binary for hex address
  binary = cache.hexToBinary(hex);
  console.log("Binary of address -> " +binary);

  // Disect binary address
  disected_bits = cache.disectBinary(binary);
  bit_selector_bits = disected_bits[0];
  tag_bits = disected_bits[1];

  // Check if hit or miss for address
  if(cache.hitOrMiss(bit_selector_bits, tag_bits) == "HIT")
    hits += 1;
  else
    misses += 1;
}

console.log("Total Hit Count = " + hits);
console.log("Total Miss Count = " + misses);

File: mt/computer_architecture_3021/assignments/6/cache.py
import math
import time

class Cache(object):
	L = 0
	K = 0
	N = 0
	Offset = 4
	set_selector_bits_length = 0
	size = 0
	tag_bits_length = 0
	cache_matrix = []

	def __init__(self, L, K, N):
		self.L = L
		self.K = K
		self.N = N
		self.size = L * K * N
		self.set_selector_bits_length = math.log(N, 2)
		self.tag_bits_length = L - self.Offset - self.set_selector_bits_length
		for i in range(0, N):
			self.cache_matrix.append([bin(i)[2:].zfill(int(self.set_selector_bits_length)),0])
			print(str(bin(i)[2:].zfill(int(self.set_selector_bits_length))))
			for j in range(0, K):
				self.cache_matrix[i].append([0,"",time.time()])


	def print_stats(self):
		print("L: " + str(self.L))
		print("K: " + str(self.K))
		print("N: " + str(self.N))
		print("Size: " + str(self.size))
		print("Offest: " + str(self.Offset))
		print("Set selector bits length: " + str(self.set_selector_bits_length))
		print("tag bits length: " + str(self.tag_bits_length))
		print("")

	def hex_to_binary(self, hex):
		binary = bin(int(hex, 16))[2:].zfill(16)
		return binary

	def disect_binary(self, binary):
		binary_minus_offset = binary[:12]
		offset = binary[12:]
		bit_selector_bits = binary_minus_offset[-int(self.set_selector_bits_length):]
		tag_bits = binary_minus_offset[:-int(self.set_selector_bits_length)]

		if self.N is 1:
			tag_bits = bit_selector_bits
			bit_selector_bits = "0"

		return bit_selector_bits, tag_bits, offset

	def find_lru(self, timestamps):
		oldest = min(timestamps)
		index = timestamps.index(oldest)
		return index

	def hit_or_miss(self, bit_selector_bits, tag_bits):

		for row in self.cache_matrix:
			if bit_selector_bits == row[0]:
				if self.K is 1:
					if row[2][0] == 0:
						row[2][1] = tag_bits
						row[2][0] = 1
						return("MISS")
					elif row[2][0] == 1:
						if row[2][1] == tag_bits:
							return("HIT")
						else:
							row[2][1] = tag_bits
							return("MISS")

				elif self.N is 1:
					timestamps = []
					for index in range(0, self.K):
						timestamps.append(row[2+index][2])
						if row[2+index][0] == 1:
							if row[2+index][1] == tag_bits:
								row[2+index][2] = time.time()
								return("HIT")

					row[1] = self.find_lru(timestamps)
					lru = row[1]

					if row[2+lru][0] == 0:
						row[2+lru][1] = tag_bits
						row[2+lru][0] = 1
						row[2+lru][2] = time.time()
						return("MISS")

					elif row[2+lru][0] == 1:
						row[2+lru][1] = tag_bits
						row[2+lru][2] = time.time()
						return("MISS")

				else:
					timestamps = []
					for index in range(0, self.K):
						timestamps.append(row[2+index][2])
						if row[2+index][0] == 1:
							if row[2+index][1] == tag_bits:
								row[2+index][2] = time.time()
								return("HIT")

					row[1] = self.find_lru(timestamps)
					lru = row[1]

					if row[2+lru][0] == 0:
						row[2+lru][1] = tag_bits
						row[2+lru][0] = 1
						row[2+lru][2] = time.time()
						return("MISS")

					elif row[2+lru][0] == 1:
						row[2+lru][1] = tag_bits
						row[2+lru][2] = time.time()
						return("MISS")

input = ["0000","0004","000c","2200","00d0","00e0","1130","0028",
		 "113c","2204","0010","0020","0004","0040","2208","0008",
		 "00a0","0004","1104","0028","000c","0084","000c","3390",
		 "00b0","1100","0028","0064","0070","00d0","0008","3394"]

myCache = Cache(16,1,8)

myCache.print_stats()

hits = 0
misses = 0
for hex in input:
	binary = myCache.hex_to_binary(hex)
	print("Binary -> " + binary)
	disected_bits = myCache.disect_binary(binary)
	bit_selector_bits = disected_bits[0]
	tag_bits = disected_bits[1]
	hit_or_miss = myCache.hit_or_miss(bit_selector_bits, tag_bits)
	if hit_or_miss == "HIT":
		hits += 1
	else:
		misses += 1
	print("0x" + hex + " " + tag_bits + " " + bit_selector_bits + " " + disected_bits[2] + " was a " + str(hit_or_miss))

print("HIT COUNT: " + str(hits))
print("MISS COUNT: " + str(misses))

# binary = myCache.hex_to_binary('0000')
# print("Binary -> " + binary)
# disected_bits = myCache.disect_binary(binary)
# bit_selector_bits = disected_bits[0]
# tag_bits = disected_bits[1]
# hit_or_miss = myCache.hit_or_miss(bit_selector_bits, tag_bits)
